File: apps/web/src/lib/money/calc.ts
import Decimal from "decimal.js";

// Konfigurasi presisi Decimal.js untuk perhitungan uang
// ROUND_HALF_UP (4) adalah pembulatan standar away-from-zero (sesuai Go shopspring/decimal)
Decimal.set({
  precision: 28,
  rounding: Decimal.ROUND_HALF_UP,
  toExpNeg: -9,
  toExpPos: 28,
});

export class RefundExceedsTotalError extends Error {
  constructor(message = "Nominal refund melebihi total transaksi yang tersisa") {
    super(message);
    this.name = "RefundExceedsTotalError";
  }
}

export interface MoneyItem {
  quantity: string | number | Decimal;
  unitPrice: string | number | Decimal;
  discount?: string | number | Decimal;
}

export interface MoneyInput {
  items: MoneyItem[];
  discount?: string | number | Decimal; // diskon keranjang
  taxRate?: string | number | Decimal; // misal 0.11 untuk 11%
}

export interface MoneyTotal {
  subtotal: Decimal;
  discountTotal: Decimal;
  taxTotal: Decimal;
  grandTotal: Decimal;
}

/**
 * calculateCart menghitung total keranjang kasir dengan paritas 100%
 * terhadap implementasi Go (services/pos-engine/internal/money/calc.go).
 */
export function calculateCart(input: MoneyInput): MoneyTotal {
  let subtotal = new Decimal(0);
  let lineDiscounts = new Decimal(0);

  for (const it of input.items || []) {
    const qty = new Decimal(it.quantity || 0);
    const unitPrice = new Decimal(it.unitPrice || 0);
    const lineDiscount = new Decimal(it.discount || 0);

    subtotal = subtotal.plus(unitPrice.times(qty));
    lineDiscounts = lineDiscounts.plus(lineDiscount);
  }

  const cartDiscount = new Decimal(input.discount || 0);
  const discountTotal = lineDiscounts.plus(cartDiscount);

  let taxable = subtotal.minus(discountTotal);
  if (taxable.isNegative()) {
    taxable = new Decimal(0);
  }

  const taxRate = new Decimal(input.taxRate || 0);
  const taxTotal = taxable.times(taxRate).toDecimalPlaces(2, Decimal.ROUND_HALF_UP);
  const grandTotal = taxable.plus(taxTotal).toDecimalPlaces(2, Decimal.ROUND_HALF_UP);

  return {
    subtotal: subtotal.toDecimalPlaces(2, Decimal.ROUND_HALF_UP),
    discountTotal: discountTotal.toDecimalPlaces(2, Decimal.ROUND_HALF_UP),
    taxTotal,
    grandTotal,
  };
}

/**
 * sumPayments menjumlahkan seluruh item pembayaran (split payments FR-22).
 */
export function sumPayments(payments: Array<string | number | Decimal>): Decimal {
  let sum = new Decimal(0);
  for (const p of payments) {
    sum = sum.plus(new Decimal(p || 0));
  }
  return sum;
}

/**
 * matchesTotal membandingkan total pembayaran terhadap grand total.
 */
export function matchesTotal(
  payments: Array<string | number | Decimal>,
  grandTotal: string | number | Decimal,
): boolean {
  return sumPayments(payments).equals(new Decimal(grandTotal));
}

/**
 * remainingRefundable menghitung sisa dana yang dapat direfund dari sebuah transaksi.
 */
export function remainingRefundable(
  grandTotal: string | number | Decimal,
  alreadyRefunded: string | number | Decimal,
  requested: string | number | Decimal,
): Decimal {
  const gTotal = new Decimal(grandTotal);
  const refunded = new Decimal(alreadyRefunded);
  const req = new Decimal(requested);

  const remaining = gTotal.minus(refunded);
  if (req.greaterThan(remaining)) {
    throw new RefundExceedsTotalError();
  }
  return remaining.minus(req);
}

File: services/pos-engine/internal/money/calc.go
// Package money menghitung total keranjang, validasi split payment, dan sisa
// refund — jalur uang di TESTING-STRATEGY.md §3A ("rumus uang wajib 100% unit
// tested"). Semuanya lewat decimal.Decimal; float64 tidak pernah dipakai
// (CLAUDE.md §5, §6.3).
package money

import "github.com/shopspring/decimal"

// Item adalah satu baris keranjang: qty × unit_price, dikurangi diskon baris.
type Item struct {
	Quantity  decimal.Decimal `json:"quantity"`
	UnitPrice decimal.Decimal `json:"unit_price"`
	Discount  decimal.Decimal `json:"discount"`
}

// Input adalah keranjang lengkap sebelum dihitung.
type Input struct {
	Items    []Item          `json:"items"`
	Discount decimal.Decimal `json:"discount"` // diskon tingkat keranjang, nominal absolut
	TaxRate  decimal.Decimal `json:"tax_rate"` // mis. 0.11 untuk PPN 11%
}

// Total adalah hasil kalkulasi, cocok dengan kolom sales_transactions
// (subtotal, discount_total, tax_total, grand_total — semua DECIMAL(14,2)).
type Total struct {
	Subtotal      decimal.Decimal `json:"subtotal"`
	DiscountTotal decimal.Decimal `json:"discount_total"`
	TaxTotal      decimal.Decimal `json:"tax_total"`
	GrandTotal    decimal.Decimal `json:"grand_total"`
}

// Calculate menghitung total keranjang.
//
// CATATAN PEMBULATAN: tidak ada aturan pembulatan yang terdokumentasi di
// manapun di repo ini (dicek 30-data/DATA-MODEL.md, 00-product/,
// docs/60-quality/TESTING-STRATEGY.md — hanya kewajiban MENGUJI kasus `.005`,
// bukan aturan hasilnya harus apa). Implementasi ini memakai round-half-up
// (decimal.Decimal.Round, "away from zero") sebagai ASUMSI, bukan keputusan
// yang sudah diriset. Sesuai CLAUDE.md §8 ("jangan mengarang angka"), ini
// perlu dikonfirmasi lewat ADR terpisah sebelum dianggap final.
func Calculate(in Input) Total {
	subtotal := decimal.Zero
	lineDiscounts := decimal.Zero
	for _, it := range in.Items {
		subtotal = subtotal.Add(it.UnitPrice.Mul(it.Quantity))
		lineDiscounts = lineDiscounts.Add(it.Discount)
	}

	discountTotal := lineDiscounts.Add(in.Discount)
	taxable := subtotal.Sub(discountTotal)
	if taxable.IsNegative() {
		taxable = decimal.Zero
	}

	taxTotal := taxable.Mul(in.TaxRate).Round(2)
	grandTotal := taxable.Add(taxTotal).Round(2)

	return Total{
		Subtotal:      subtotal.Round(2),
		DiscountTotal: discountTotal.Round(2),
		TaxTotal:      taxTotal,
		GrandTotal:    grandTotal,
	}
}

// SumPayments menjumlahkan seluruh baris pembayaran (FR-22 split payment —
// mis. Rp 20.000 tunai + Rp 30.000 QRIS dalam satu transaksi).
func SumPayments(payments []decimal.Decimal) decimal.Decimal {
	sum := decimal.Zero
	for _, p := range payments {
		sum = sum.Add(p)
	}
	return sum
}

// MatchesTotal membandingkan jumlah pembayaran dengan grand_total. Dipakai
// checkout handler untuk menegakkan PAYMENT_AMOUNT_MISMATCH (ERROR-CATALOG §D).
func MatchesTotal(payments []decimal.Decimal, grandTotal decimal.Decimal) bool {
	return SumPayments(payments).Equal(grandTotal)
}

// RemainingRefundable menghitung sisa yang boleh di-refund dari sebuah
// transaksi. Mengembalikan error bila permintaan melebihi sisa
// (REFUND_EXCEEDS_TOTAL, ERROR-CATALOG §B).
func RemainingRefundable(grandTotal, alreadyRefunded, requested decimal.Decimal) (decimal.Decimal, error) {
	remaining := grandTotal.Sub(alreadyRefunded)
	if requested.GreaterThan(remaining) {
		return remaining, ErrRefundExceedsTotal
	}
	return remaining.Sub(requested), nil
}
